buffer/src/main/java/io/netty/buffer/UnsafeByteBufUtil.java

```diff
@@ -538,34 +538,48 @@ static void setBytes(AbstractByteBuf buf, long addr, int index, byte[] src, int
     }
 
     static void setBytes(AbstractByteBuf buf, long addr, int index, ByteBuffer src) {
-        buf.checkIndex(index, src.remaining());
-
-        int length = src.remaining();
+        final int length = src.remaining();
         if (length == 0) {
             return;
         }
 
         if (src.isDirect()) {
+            buf.checkIndex(index, length);
             // Copy from direct memory
             long srcAddress = PlatformDependent.directBufferAddress(src);
-            PlatformDependent.copyMemory(srcAddress + src.position(), addr, src.remaining());
+            PlatformDependent.copyMemory(srcAddress + src.position(), addr, length);
             src.position(src.position() + length);
         } else if (src.hasArray()) {
+            buf.checkIndex(index, length);
             // Copy from array
             PlatformDependent.copyMemory(src.array(), src.arrayOffset() + src.position(), addr, length);
             src.position(src.position() + length);
         } else {
-            ByteBuf tmpBuf = buf.alloc().heapBuffer(length);
-            try {
-                byte[] tmp = tmpBuf.array();
-                src.get(tmp, tmpBuf.arrayOffset(), length); // moves the src position too
-                PlatformDependent.copyMemory(tmp, tmpBuf.arrayOffset(), addr, length);
-            } finally {
-                tmpBuf.release();
+            if (length < 8) {
+                setSingleBytes(buf, addr, index, src, length);
+            } else {
+                //no need to checkIndex: internalNioBuffer is already taking care of it
+                assert buf.nioBufferCount() == 1;
+                final ByteBuffer internalBuffer = buf.internalNioBuffer(index, length);
+                internalBuffer.put(src);
```

Comment:
Update src.position?

Reply:
AFAIK `ByteBuffer::put(ByteBuffer)` will do it

Reply:
The ByteBuffer has a copy of the indices which will be modified, but that won't be reflected in the ByteBuf.

Reply:
And that's seems ok to me: `buf` indices are not supposed to be modified

Reply:
Oh, you're right. I had it inverted in my mind; we're reading from a ByteBuffer, not a ByteBuf.

Reply:
np ^^ I often cross my eyes on `ByteBufer::put` instead: why it isn't named as `ByteBufer::readFrom` or something more human readable? :(